exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -307,3 +315,246 @@ func (r *MachinePoolReconciler) reconcileInfrastructure(ctx context.Context, clu
 
 	return ctrl.Result{}, nil
 }
+
+// reconcileMachinePoolMachines reconciles MachinePool Machines associated with a MachinePool.
+func (r *MachinePoolReconciler) reconcileMachinePoolMachines(ctx context.Context, mp *expv1.MachinePool, infraMachinePool *unstructured.Unstructured) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	var noKind bool
+	var noSelector bool
+
+	var infraMachineKind string
+	if err := util.UnstructuredUnmarshalField(infraMachinePool, &infraMachineKind, "status", "infrastructureMachineKind"); err != nil {
+		if errors.Is(err, util.ErrUnstructuredFieldNotFound) {
+			noKind = true
+		} else {
+			return errors.Wrapf(err, "failed to retrieve infraMachineKind from infrastructure provider for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+		}
+	}
+
+	var infraMachineSelector metav1.LabelSelector
```

Comment:
I think that would be reasonable as long as it's clear in documentation. Would we want to change it at this stage?

Reply:
@CecileRobertMichon @mboersma  opinions?

Reply:
I think Fabrizio means that we could use `infraMachineKind` to be the sole indicator of whether or not MachinePool Machines are supported. Currently, MachinePool Machines are only supported if both fields are set.

The value of having `infraMachineSelector` would be that the infra provider can specify their own labels they want to use. If we go with Fabrizio's suggestion, we would essentially hard code a selector and require that all infraMachines have the `MachinePoolNameLabel` and `ClusterNameLabel`.

Reply:
> require that all infraMachines have the MachinePoolNameLabel and ClusterNameLabel.

as long as that contract is well documented (both in the migration doc and in the infra provider contract) I think that would be fine to start with. We can always add support for a custom selector later if a provider has a use case for it. It would make the UX more consistent across providers too.

Reply:
I just realized my comment wasn't clear above, edited to clarify. "start with" should have been "start without it"

Reply:
I think that's a reasonable change, though it would be a departure from the proposal. Do you think it's too late to introduce this change? @fabriziopandini @sbueringer

Reply:
I think it's fine to introduce this change now. Looking at the comments above Fabrizio and Cecile seem to agree